test/lib/jdk/test/lib/Asserts.java

```diff
@@ -552,6 +552,42 @@ public static void assertStringsEqual(String str1, String str2,
         }
     }
 
+    /**
+     * A functional interface for executing tests in assertThrownException
+     */
+    @FunctionalInterface
+    public interface TestMethod {
+        void execute() throws Exception;
+    }
+
+    /**
+     * Asserts that the given exception (or a subclass of it) is thrown when
+     * executing the test method.
+     *
+     * If the test method throws the correct exception, the exception is returned
+     * to the caller for additional validation e.g., comparing the exception
+     * message.
+     *
+     * @param expected The expected exception
+     * @param testMethod The code to execute that should throw the exception
+     * @return The thrown exception.
+     */
+    public static Exception assertThrownException(Class<? extends Exception> expected, TestMethod testMethod) {
+        try {
+            testMethod.execute();
+            fail("No exception was thrown. Expected: " + expected.getName());
+        } catch (Exception exc) {
+            if (expected.isAssignableFrom(exc.getClass())) {
+                return exc;
+            } else {
+                fail("An unexpected exception was thrown. Expected: "
+                        + expected.getName() + " Got: " + exc.getClass().getName());
+            }
+        }
+        // fail() throws a RuntimeException so this is unreachable.
+        return null;
```

Comment:
Actually, this is a mistake. If I call `fail()` immediately after the test method executes (because an exception wasn't thrown), the RuntimeException which will promptly be caught in my try/catch and that isn't what we want. I'm moving that invocation to the end which looks and works a lot better.

Reply:
Ah, right. Throwing unconditionally at the end of the method is the right thing. Good catch.